ch_util/cal_utils.py

```diff
@@ -2161,3 +2166,315 @@ def _dec_to_el(dec):
+    n_bad_times = np.sum(~np.isfinite(reftime[~known_bad_times]))
+    if n_bad_times > 0:
+        msg = (
+            "{0} out of {1} time entries don't have a reference calibration time "
+            + "without an identifiable cause. Cannot correct gains for those entries."
+        )
+        logger.warning(msg.format(n_bad_times, ntimes))
+
+    # Bundle result in dictionary
+    result = {
+        "reftime": reftime,
+        "reftime_prev": reftime_prev,
+        "interp_start": interp_start,
+        "interp_stop": interp_stop,
+    }
+
+    return result
+
+
+def get_reference_times_dataset_id(
+    times: np.ndarray,
+    dataset_ids: np.ndarray,
+    logger: Optional[logging.Logger] = None,
+) -> Dict[str, np.ndarray]:
+    """Calculate the relevant calibration reference times from the dataset IDs.
+
+    .. warning::
+        Dataset IDs before 2020/10/10 are corrupt so this routine won't work.
+
+    Parameters
+    ----------
+    times
+        Unix times of data points to be calibrated as floats.
+    dataset_ids
+        The dataset IDs as an array of strings.
+    logger
+        A logging object to use for messages. If not provided, use a module level
+        logger.
+
+    Returns
+    -------
+    reftime_result
+        A dictionary containing the results. See `get_reference_times_file` for a
+        description of the contents.
+    """
+    if logger is None:
+        logger = logging.getLogger(__name__)
+
+    # Dataset IDs before this date are untrustworthy
+    ds_start = ephemeris.unix_to_datetime(datetime(2020, 10, 10))
+    if (times < ds_start).any():
+        raise ValueError(
+            "Dataset IDs before 2020/10/10 are corrupt, so this method won't work. "
+            f"You passed in a time as early as {ctime.unix_to_datetime(times.min())}."
+        )
+
+    # The CHIME calibration sources
+    _source_dict = {
+        "cyga": ephemeris.CygA,
+        "casa": ephemeris.CasA,
+        "taua": ephemeris.TauA,
+        "vira": ephemeris.VirA,
+    }
+
+    # Get the set of gain IDs for each time stamp
+    gain_ids = state_id_of_type(dataset_ids, "gains")
+    collapsed_ids = unique_unmasked_entry(gain_ids, axis=0)
+    unique_gains_ids = np.unique(collapsed_ids.compressed())
+
+    gain_info_dict = {}
+
+    # For each gain update extract all the relevant information
+    for state_id in unique_gains_ids:
+
+        d = {}
+        gain_info_dict[state_id] = d
+
+        # Extract the update ID
+        update_id = ds.DatasetState.from_id(state_id).data["data"]["update_id"]
+
+        # Parse the ID for the required information
+        split_id = update_id.split("_")
+        # After restart we sometimes have only a timing update without a source
+        # reference. These aren't valid for our purposes here, and can be distinguished
+        # at the update_id doesn't contain source information, and is thus shorter
+        d["valid"] = len(split_id) > 3
+
```

Comment:
```suggestion
d["interpolated"] = "transition" in split_id
```
Suggest having an `interpolated` entry for all non-null ids (assumed below).